### DISCUSSION TITLE
DevTest Lab i podłączenie repozytorium Azure DevOps

### DISCUSSION THREAD

**Discussion by ghost** [AzureCommunityPL/group#311] [Pomoc]
Witam,
Utworzyłem w Azure usługę Devtest Lab i chciałbym do niej podłączyć zewnętrzne repo w Azure DevOps aby korzystać z Artifacts i ARMTemplates. 
Z poziomu DevOps jest projekt X do którego mam dostęp contributora. W ramach projektu mam repo Y. Utworzyłem testowo Personal Access Token - PAT typu full access a także typu Code(read).
Od strony Azure DevTest Labs przechodzę do Configuration and policies > Repositories >Add.
Wprowadzam dane, w tym URL (url jest poprawny bo rozwiązuje mi nazwę) i PAT. Niestety otrzymuję po chwili błąd "GeneralError: GeneralError". testowałem to na dwóch PATach.
![repo_devops](https://user-images.githubusercontent.com/72250890/123365798-4d525d00-d577-11eb-8cf1-816a4f4f86bd.jpg)

Czy ktoś przerabiał powyższy temat? Robię to pierwszy raz a nie znalazłem nic w necie. PAT dodałem wg. oficjalnego doca od MS, korzystałem również z https://docs.microsoft.com/pl-pl/azure/devtest-labs/add-artifact-repository.

Jak DevtestLab weryfikuje PAT ?
Jakie uprawnienia powinienem zweryfikować?

Wszelkie informacje mile widziane.

Zapisanie tego PAT w keyvault w danym labie w DevTestLab niestety tez nic nie daje. Wówczas za pomocą skryptu powershell korzystam z:

**DevLabs-AddRepository.json**
{
    "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
    "contentVersion": "1.0.0.0",
    "parameters": {
        "artifactRepoSecurityToken": {
            "type": "securestring"
        },
        "artifactRepositoryDisplayName": {
            "type": "string",
            "defaultValue": "Nazwa repo"
        },
        "artifactRepoUri": {
            "type": "string",
            "defaultValue": "https://xxx.visualstudio.com/yyy/_git/zzz"
        },
        "artifactRepoBranch": {
            "type": "string",
            "defaultValue": "master"
        },
        "artifactRepoFolder": {
            "type": "string",
            "defaultValue": "/Artifacts"
        },
        "ARMRepoFolder": {
            "type": "string",
            "defaultValue": "/ARMTemplates"
        },
        "artifactRepoType": {
            "type": "string",
            "defaultValue": "VsoGit",
            "allowedValues": [
                "VsoGit",
                "GitHub"
            ]
        }
    },
    "variables": {},
    "resources": [
        {
            "apiVersion": "2018-10-15-preview",
            "type": "Microsoft.DevTestLab/labs/artifactSources",
            "name": "[concat(trim(resourceGroup().name), '/',parameters('artifactRepositoryDisplayName'))]",
            "location": "[resourceGroup().location]",
            "properties": {
                "uri": "[parameters('artifactRepoUri')]",
                "folderPath": "[parameters('artifactRepoFolder')]",
                "armTemplateFolderPath": "[parameters('ARMRepoFolder')]",
                "branchRef": "[parameters('artifactRepoBranch')]",
                "displayName": "[parameters('artifactRepositoryDisplayName')]",
                "securityToken": "[parameters('artifactRepoSecurityToken')]",
                "sourceType": "[parameters('artifactRepoType')]",
                "status": "Enabled"
            }
        }
    ],
    "outputs": {
    }
}

**DevLabs-AddRepository.parameters.json**
{
    "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentParameters.json#",
    "contentVersion": "1.0.0.0",
    "parameters": {
        "artifactRepoSecurityToken": {
            "reference": {
                "keyVault": {
                    "id": "/subscriptions/id suby/resourceGroups/nazwa rg gdzie jest keyvault/providers/Microsoft.KeyVault/vaults/nazwakeyvault"
                },
                "secretName": "nazwa security tokena"
            }
        },
        "artifactRepositoryDisplayName": {
            "value": "nazwa repo"
        },
        "artifactRepoUri": {
            "value": "https://xxx.visualstudio.com/yyy/_git/zzz"
        },
        "artifactRepoBranch": {
            "value": "master"
        },
        "artifactRepoFolder": {
            "value": "/Artifacts"
        },
        "ARMRepoFolder": {
            "value": "/ARMTemplates"
        },
        "artifactRepoType": {
            "value": "VsoGit"
        }
    }
}

Deployment niestety się failuje ze statusem Bad Request,

**Comment by smajchrz** [comment]
Może głupie pytanie. Czy foldery w repo masz już utworzone?
Spróbuj również użyć adresu repo w formacie ```https://dev.azure.com/<ORG>/<PROJECT>/_git/<REPO>```

Jeśli chodzi o PAT to uprawnienia Code (Read) będą wystarczające.

**Reply by ghost** [reply]
Tak foldery są juz utworzone. Ja przejąłem ten temat po innej osobie, może to jest jednak kwestia moich uprawnień, chociaż mam contributor w projekcie.

Adres w formacie https://org.visualstudio.com/IT/_git/DevXXX zadziałał do innych DevTest Labs utworzonych poprzednio.

**Comment by sbugalski** [comment]
Cześć, u mnie przeszło z oficjalnym repo https://github.com/Azure/azure-devtestlab sklonowanym do AzDO

Jeżeli u Ciebie nie przechodzi z tym repo zerknij do Activity Log, a jak nie to pozostaje support.

![image](https://user-images.githubusercontent.com/45152735/124356849-a3c73780-dc18-11eb-859e-8c4be2481c7c.png)
